Allow keyboard smooth scrolling to take place in overflow:scroll regions

danae404 opened the pull request:
#### 30609c3f1dc679ab91ff12f87ba0790487f3f02a
<pre>
Allow keyboard smooth scrolling to take place in overflow:scroll regions
<a href="https://bugs.webkit.org/show_bug.cgi?id=228158">https://bugs.webkit.org/show_bug.cgi?id=228158</a>
rdar://80911829

Reviewed by Tim Horton.

Test: LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll.html

Right now, keyboard smooth scrolling can only scroll the entire page and
cannot scroll overflow:scroll regions. To fix this, rather than always
obtaining ScrollingAnimator from m_frame-&gt;view(), which will only ever
scroll the whole page, the code should get ScrollingAnimator from
m_frame.document()-&gt;focusedElement() or m_mousePressNode if either one
of them is non-null, which indicates that a sub-scrollable region is
focused.

* LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(Webcore::EventHandler::keyboardScrollingAnimatorForFocusedNode):
(WebCore::EventHandler::stopKeyboardScrolling):
(WebCore::EventHandler::startKeyboardScrolling):
* Source/WebCore/page/EventHandler.h:
</pre>

hortont424 commented on LayoutTests/fast/scrolling/keyboard-scrolling-subframe-expected.txt:
```diff
@@ -0,0 +1,2 @@
+Successful.
```
Your whole commit message is about "subframes", but the change is about `overflow: scroll` style, which is not a subframe. "sub/child scrollable regions" or "overflow scroll regions" or something might be better wording. (Subframes would be like <iframe>, where you literally have multiple WebCore Frame objects)

hortont424 commented on LayoutTests/fast/scrolling/keyboard-scrolling-subframe.html:
```diff
@@ -0,0 +1,67 @@
+                return;
+
+            let scroller = document.getElementById("scroller");
+
```
Style nit: lots and lots of extraneous newlines in this function

hortont424 commented on LayoutTests/fast/scrolling/keyboard-scrolling-subframe.html:
```diff
@@ -0,0 +1,67 @@
+
+            document.addEventListener("scroll", documentScrolled);
+
+            await UIHelper.doubleClickAt(10, 10);
```
Why a double click? Single click should be plenty to put the selection in here.

hortont424 commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4339,28 +4339,37 @@ void EventHandler::defaultBackspaceEventHandler(KeyboardEvent& event)
-    Ref protectedFrame = m_frame;
-    auto* view = m_frame.view();
-    if (!view)
+    auto* node = m_mousePressNode.get();
```
Can we factor "mouse press node -> enclosing scrollable area" out into a function here somewhere? Because I think you are adding the second and third (at least?) copy of these.

Also, do the other clients /just/ look at mousePressNode? I would expect to be able to move the focus with the keyboard too, so I'm a bit surprised. What does the existing code do?

danae404 replied:
> Can we factor "mouse press node -> enclosing scrollable area" out into a function here somewhere? Because I think you are adding the second and third (at least?) copy of these.

I'm having trouble finding where else in EventHandler.cpp there is similar code that can be factored out.


> Also, do the other clients /just/ look at mousePressNode? I would expect to be able to move the focus with the keyboard too, so I'm a bit surprised. What does the existing code do?

I see that from the non-smooth code, EventHandler::scrollOverflow first tries to get the node from m_frame.document()->focusedElement(), and then if that fails tries m_mousePressNode.get(). I _think_ this would address that issue, so I included that change in the new patch, but I'm not certain.

danae404 committed:
Allow keyboard smooth scrolling to take place in overflow:scroll regions
https://bugs.webkit.org/show_bug.cgi?id=228158
rdar://80911829

Reviewed by Tim Horton.

Test: LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll.html

Right now, keyboard smooth scrolling can only scroll the entire page and
cannot scroll overflow:scroll regions. To fix this, rather than always
obtaining ScrollingAnimator from m_frame->view(), which will only ever
scroll the whole page, the code should get ScrollingAnimator from
m_frame.document()->focusedElement() or m_mousePressNode if either one
of them is non-null, which indicates that a sub-scrollable region is
focused.

* LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/keyboard-scrolling-overflow-scroll.html: Added.
* Source/WebCore/page/EventHandler.cpp:
(Webcore::EventHandler::keyboardScrollingAnimatorForFocusedNode):
(WebCore::EventHandler::stopKeyboardScrolling):
(WebCore::EventHandler::startKeyboardScrolling):
* Source/WebCore/page/EventHandler.h:

megangardner commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -29,6 +29,7 @@
 #include "DragActions.h"
 #include "FocusDirection.h"
 #include "HitTestRequest.h"
+#include "KeyboardScrollingAnimator.h"
```
Don't move this to the .h add 
class KeyboardScrollingAnimator; below
